web/src/typeahead_helper.ts

```diff
@@ -1034,15 +1034,45 @@ export function query_matches_person(
     query: string,
     person: UserPillData | UserOrMentionPillData,
 ): boolean {
-    if (typeahead.query_matches_string_in_order(query, person.user.full_name, " ")) {
+    if (
+        person.type === "broadcast" &&
+        typeahead.query_matches_string_in_order(query, person.user.full_name, " ")
+    ) {
         return true;
     }
-    if (person.type === "user" && Boolean(person.user.delivery_email)) {
-        return typeahead.query_matches_string_in_order(
-            query,
-            people.get_visible_email(person.user),
-            " ",
-        );
+    if (person.type === "user") {
+        query = query.toLowerCase();
+        const full_name = person.user.full_name.toLowerCase();
+
+        // We use *, because we want to start caching from empty query.
+        const is_query_ascii = /^[a-z]*$/.test(query);
+        // Ignore diacritics from `full_name` only if `query` contains only ASCII characters.
+        // This ensures diacritic-sensitive queries match accurately.
+        if (is_query_ascii) {
+            // Reuse removed diacritics version of the `full_name` if present
+            // else compute and cache it.
+            person.user.name_with_diacritics_removed ??= typeahead.remove_diacritics(full_name);
```

Comment:
This conflicts with the meaning of `name_with_diacritics_removed` in `people.build_termlet_matcher`, where it stores `remove_diacritics(user.full_name)`, not `remove_diacritics(user.full_name.toLowerCase())`.

(Why do we have two separate code paths for this, anyway?)